### PR TITLE
Round restart

### PR DESCRIPTION
**Pull request by kellpossible**
Currently a work in progress:

TODO:

- [x] Handle situation where there are no replacement contributors and there are some remaining contributors after the drop of a contributor.
- [x] Update integration test to support starting contributors while a round is in progress, so the situation where there are no contributors left in the current round after a drop can be tested. https://github.com/AleoHQ/aleo-setup-integration-test/pull/4
- [x] Handle situation where there are no contributors left.
- [x] #286 Manual round restart (which should possibly remove everyone and take people from the queue again? Or have an option for doing so).
- [x] ~Check what happens when a verifier drops (do the remaining verifiers pick up the dropped verifier's tasks successfully?).~ We can do this in #305 
- [x] ~Handle situation where there are no verifiers left in the current round. Or at least document this in an issue.~ We can do this in #305 
- [x] ~Handle situation where there are no replacement contributors left~ (or at least document it for future implementation of the replacement contributor logic). #309

A number of refactors in this PR are related to #267

### CODE REVIEW

**Comment by apruden2008** [comment]
@kellpossible are we separating the unfinished items into new issues?

**Comment by kellpossible** [comment]
@apruden2008 I think so yes, I'll do that today, also I will properly report the problems that I encountered while testing the manual round restart in new issues

File: phase1-coordinator/src/storage/storage.rs
Comment (apruden2008) [inline phase1-coordinator/src/storage/storage.rs:321]: I will open an issue for this

**Review by julesdesmit — APPROVED** [review]
LGTM!

**Comment by kellpossible** [comment]
Pending a second review, hopefully this PR is good to go

File: phase1-coordinator/src/coordinator_state.rs
Comment (ibaryshnikov) [inline phase1-coordinator/src/coordinator_state.rs:1025]: It looks like one of the key parts of the change, is it? To reset the state of all contributors who is still in the round to the initial state similar to what we have on round start.

Comment (kellpossible) [reply phase1-coordinator/src/coordinator_state.rs:1025]: yep that's right

File: phase1-coordinator/src/coordinator_state.rs
Comment (ibaryshnikov) [inline phase1-coordinator/src/coordinator_state.rs:2375]: is it enough to have `self.current_verifiers.is_empty()`? Can you tell us more about the finished verifiers check here?

File: phase1-coordinator/src/objects/round.rs
Comment (ibaryshnikov) [inline phase1-coordinator/src/objects/round.rs:798]: I hope we will remove "_unsafe" and "try_" from the function names one day. In case of `remove_something_unsafe` it's misleading because it's a safe function. And `try_do_something() -> Result<...>` doesn't need to have "try_" because it returns `Result` and it's clear that it can fail. And here the name is justified by the similarity to other `remove_*_unsafe` functions, which is ok for now.

**Review by ibaryshnikov — APPROVED** [review]

File: phase1-coordinator/src/objects/round.rs
Comment (kellpossible) [reply phase1-coordinator/src/objects/round.rs:798]: I agree, I was just following convention in this case. I think perhaps `try_` doesn't quite convey the intended meaning. The function may not fail with a result, but my guess is the original authors were trying to say that calling this function might be dangerous in the wrong contexts (without being "unsafe" according to the strict Rust definition).

File: phase1-coordinator/src/coordinator_state.rs
Comment (kellpossible) [reply phase1-coordinator/src/coordinator_state.rs:2375]: The verifiers in the round get moved from the `current_verifiers` list to the `finished_verifiers` list when all their tasks are complete. If all the verifiers are considered complete for whatever reason, then the round will still be reset.

Comment (kellpossible) [reply phase1-coordinator/src/coordinator_state.rs:2375]: at least that's how I was understanding it

Comment (ibaryshnikov) [reply phase1-coordinator/src/coordinator_state.rs:2375]: Is this case possible: one verifier finished, another one dropped?. This check will be produce therefore.

Comment (kellpossible) [reply phase1-coordinator/src/coordinator_state.rs:2375]: Thanks for raising questions about this line, I think it deserves reconsideration.

Comment (kellpossible) [reply phase1-coordinator/src/coordinator_state.rs:2375]: I referenced it in #305

Comment (kellpossible) [reply phase1-coordinator/src/coordinator_state.rs:2375]: Yes it will produce false. Under the assumption that the remaining verifiers would be able to pick up the slack, but perhaps this isn't working yet and will be fixed in #305

**Pull request merged**
